timefold-solver/src/main/java/com/baeldung/timefoldsolver/Shift.java

```diff
@@ -0,0 +1,57 @@
+package com.baeldung.timefoldsolver;
+
+import java.time.LocalDateTime;
+
+import ai.timefold.solver.core.api.domain.entity.PlanningEntity;
+import ai.timefold.solver.core.api.domain.variable.PlanningVariable;
+
+@PlanningEntity
+public class Shift {
+
+    private LocalDateTime start;
+    private LocalDateTime end;
+    private String requiredSkill;
+
+    @PlanningVariable
+    private Employee employee;
+
+    // A no-arg constructor is required for @PlanningEntity annotated classes
```

Comment:
Out of curiousity, why is it that @PlanningEntity classes need a no-arg constructor, but @PlanningVariable types e.g. Employee seem to be fine without one?

Reply:
Because internally Timefold Solver needs to "planning clone" the entity and solution classes to keep track of the best solution (so far) while the working solution changes to find a better one.
The other classes (Employee) don't change during planning, so there is no need to do such magic on them.

Reply:
PS: you might be able to use bytecode manipulation tools such as https://bytebuddy.net/ to remove the need for a no-arg constructor. Out of the scope of this article, but could help make your user facing APIs a bit more concise.